src/http/request.ts

```diff
@@ -27,8 +27,39 @@ const printWarnings = {
 }
 
 /** @internal */
-export function defineHttpRequest(envMiddleware: Middlewares): HttpRequest {
+export function defineHttpRequest(
+  envMiddleware: Middlewares,
+  {maxRetries, retryDelay}: {maxRetries: number; retryDelay?: (attemptNumber: number) => number}
+): HttpRequest {
   const request = getIt([
+    retry({
+      retryDelay: retryDelay as () => number, // This option is typed incorrectly in get-it.
+      maxRetries,
+      shouldRetry(err, attempt, options) {
+        // By default `retry.shouldRetry` doesn't retry on server errors so we add our own logic.
+
+        const isSafe = options.method === 'GET' || options.method === 'HEAD'
+        const isQuery = options.uri.startsWith('/data/query')
+        const shouldRetry =
+          err.response &&
+          (err.response.statusCode === 429 ||
+            err.response.statusCode === 502 ||
+            err.response.statusCode === 503)
```

Comment:
We're already using `is-retry-allowed` in get-it which I believe should support this, but it's not applied to POST. Maybe the "proper" way to support this is to support `options.isSafe = true` (or maybe `options.isIdempotent = true` since we can actually also retry these?), and then have the default `shouldRetry` check for this instead?

```js
function shouldRetry(…) {
  const isSafe = options.isSafe || options.method === 'GET' || options.method === 'HEAD' // remain backwards compatible
  if (!isSafe) return false
}
```